### PR TITLE
Feat/profile/integrate vm editprofile

### PR DESCRIPTION
**Pull request by Harrish92**
# Title
Integrate ViewModel into EditProfileScreen

## Description
This PR introduces the integration of `UserViewModel` into `EditProfile` to handle loading and saving user data. It closes issue #163 .

## Changes
- Implemented `UserViewModel` into `EditProfileScreen` to use for loading user data.
-  Changed variables to manage user profile data.
- Implemented `UserViewModel` into `EditProfileScreen` to use for saving user data.
- Updated validation logic to make it similar to `CreateProfile`.

#### Modified
- `app/src/main/java/com/android/periodpals/ui/profile/EditProfile.kt`

## Testing
- Adapated `EditProfileTest` to the `UserViewModel`

### CODE REVIEW

File: app/src/main/java/com/android/periodpals/ui/profile/EditProfile.kt [hunk 1]
```diff
@@ -60,35 +75,35 @@ private const val SCREEN_TITLE = "Edit Your Profile"
...
 @Composable
-fun EditProfileScreen(navigationActions: NavigationActions) {
-  // TODO: State variables, to replace it with the real data when the implementing profile VM
-  var name by remember { mutableStateOf("Emilia Jones") }
-  var dob by remember { mutableStateOf("20/01/2001") }
-  var description by remember {
-    mutableStateOf(
-        "Hello guys :) I’m Emilia, I’m a student " +
-            "at EPFL and I’m here to participate and contribute to this amazing community !")
+fun EditProfileScreen(userViewModel: UserViewModel, navigationActions: NavigationActions) {
```
Comment (agonzalez-r) [inline app/src/main/java/com/android/periodpals/ui/profile/EditProfile.kt:80]: Please update documentation according to the changes (add `@param userViewModel`)

File: app/src/main/java/com/android/periodpals/ui/profile/EditProfile.kt [hunk 2]
```diff
@@ -158,19 +173,65 @@ fun EditProfileScreen(navigationActions: NavigationActions) {
...
+    context: Context,
+    userViewModel: UserViewModel,
+    userState: State<User?>,
+    navigationActions: NavigationActions,
+) {
+  val errorMessage = validateFields(name, age, description)
+  if (errorMessage != null) {
+    Log.d(LOG_TAG, "$LOG_FAILURE: $errorMessage")
+    Toast.makeText(context, errorMessage, Toast.LENGTH_SHORT).show()
+    return
```
Comment (agonzalez-r) [inline app/src/main/java/com/android/periodpals/ui/profile/EditProfile.kt:218]: I think you can remove this `return`

File: app/src/main/java/com/android/periodpals/ui/profile/EditProfile.kt [hunk 3]
```diff
@@ -60,35 +75,35 @@ private const val SCREEN_TITLE = "Edit Your Profile"
...
+
+  val context = LocalContext.current
+  userViewModel.loadUser()
+  val userState = userViewModel.user
+  if (userState.value == null) {
+    Log.d(TAG, "User data is null")
+    Toast.makeText(context, "Error loading your data! Try again later.", Toast.LENGTH_SHORT).show()
   }
 
+  var name by remember { mutableStateOf(userState.value?.name ?: DEFAULT_NAME) }
```
Comment (agonzalez-r) [inline app/src/main/java/com/android/periodpals/ui/profile/EditProfile.kt:90]: I would remove the default name here, since if you don't have a name/dob/description it shouldn't say "Error loading ..."

**Review by agonzalez-r — CHANGES_REQUESTED** [review]
I'm getting a bug when testing manually: when I modify the profile and click save the changes are not always reflected in the `ProfileScreen`. I see that you didn't touch the code on that screen. Please verify that the data is properly fetched from that screen when there are changes.

**Comment by Harrish92** [comment]
> I'm getting a bug when testing manually: when I modify the profile and click save the changes are not always reflected in the `ProfileScreen`.

@agonzalez-r 
An RLS policy was added to allow any modification of the profile by the user. Any changes should be reflected on the `ProfileScreen` now.

**Review by agonzalez-r — APPROVED** [review]
Bug is fixed now! Thank you for the changes. 
LGTM

File: app/src/main/java/com/android/periodpals/ui/profile/EditProfile.kt [hunk 4]
```diff
@@ -158,19 +169,65 @@ fun EditProfileScreen(navigationActions: NavigationActions) {
...
+  if (userState.value == null) {
+    Log.d(LOG_TAG, LOG_FAILURE)
+    Toast.makeText(context, TOAST_FAILURE, Toast.LENGTH_SHORT).show()
+    return
+  }
+
+  Log.d(LOG_TAG, LOG_SUCCESS)
+  Toast.makeText(context, TOAST_SUCCESS, Toast.LENGTH_SHORT).show()
+  navigationActions.navigateTo(Screen.PROFILE)
+}
```
Comment (francelu) [inline app/src/main/java/com/android/periodpals/ui/profile/EditProfile.kt:229]: This is a bit redundant with the `CreateProfileScreen` logic. Could you maybe try to extract the `attemptSaveUserData into the `ProfileSaveButton` logic ?  

Below, the last helper function `validateFields()` is an exact duplicate of the one in `CreateProfileScreen` too. Maybe put its logic in the shared profile component `ProfileSaveButton`.  

Maybe it isn't best practice, but I think having the `validateDate` and `validateFields` in `ProfileComponents.kt` would be a good idea since it's shared among profile screens.

File: app/src/main/java/com/android/periodpals/MainActivity.kt
Comment (francelu) [inline app/src/main/java/com/android/periodpals/MainActivity.kt:1]: Consistent with existing implementation. Good job !

File: app/src/androidTest/java/com/android/periodpals/ui/profile/EditProfileTest.kt [hunk 5]
```diff
@@ -31,13 +34,27 @@ import org.mockito.kotlin.never
...
   fun setUp() {
     navigationActions = mock(NavigationActions::class.java)
+    userViewModel = mock(UserViewModel::class.java)
+
     `when`(navigationActions.currentRoute()).thenReturn(Route.PROFILE)
-    composeTestRule.setContent { EditProfileScreen(navigationActions) }
+    `when`(userViewModel.user).thenReturn(userState)
+
+    composeTestRule.setContent { EditProfileScreen(userViewModel, navigationActions) }
   }
```
Comment (francelu) [inline app/src/androidTest/java/com/android/periodpals/ui/profile/EditProfileTest.kt:58]: - Have a look at the more complete similar tests in `CreateProfileTest.kt`. It check also the cases with userVM failure in `createValidProfileVMFailure()`.
- You'll need to adapt the `setUp()` as it's done in the previously mentioned file.
- Each test will also need to be adapted with "when" and `setContent()` as done in that file too.

Else, good implementation of the VM into the the tests !

File: app/src/androidTest/java/com/android/periodpals/ui/profile/EditProfileTest.kt
Comment (francelu) [inline app/src/androidTest/java/com/android/periodpals/ui/profile/EditProfileTest.kt:1]: Other after thought, you did not do tests as in `CreateProfileTest.kt` about the date validation. I'm wondering whether it would be meaningful to do it again as we are calling the exact same function itself.

**Review by francelu — APPROVED** [review]
## Summary
Overall good implementations of profile's VM into the edit profile screen !

## Important
#### Code Quality
- Code is readable, easy and mostly consistent with existing implementation of profile VM into V.
- Little nitpick about code duplication (see in-line comment in `EditProfileScreen`).

#### Functionality
- Edit profile works well ! Good job :)

#### Performance
- Unchanged.

#### Security
- Unchanged.

## Testing
- A few edge cases could be tested, see comments.

## Documentation
- You did the necessary changes, beside what @agonzalez-r pointed out.

## Suggestions
- See comments.

## Steps before PR **merged**
- Update tests with `editValidProfileVMFailure()` and mentioned changes in comments.
- Would be good to have in `ProfileSaveButton()` the logic of `attemptSaveUserData()` and `validateFields()` as it would reduce code duplication with `CreateProfileScreen`'s code.
- Minor nitpick: extract `validateDate()` into `ProfileComponents` or `ProfileSaveButton` too.
- Answer to my comments :)

File: app/src/main/java/com/android/periodpals/ui/profile/EditProfile.kt (hunk 4 above)
Comment (Harrish92) [reply app/src/main/java/com/android/periodpals/ui/profile/EditProfile.kt:229]: Since `attemptSaveUserData`, `validateFields` and `validateDate` are validation logic functions and not part of the UI, it would be more suitable to create a helper class for user profiles.

File: app/src/androidTest/java/com/android/periodpals/ui/profile/EditProfileTest.kt
Comment (Harrish92) [reply app/src/androidTest/java/com/android/periodpals/ui/profile/EditProfileTest.kt:1]: A helper class could help reduce the duplicated codes for the tests. This could be done in another issue.

Comment (Harrish92) [reply app/src/androidTest/java/com/android/periodpals/ui/profile/EditProfileTest.kt:1]: Adding these tests in `EditProfileTest` would be redundant. Therefore, it is not changed
